src/dll/edbgEcmdDll.C

```diff
@@ -216,108 +162,59 @@ static uint32_t findChipUnitType(const ecmdChipTarget &i_target, uint64_t i_addr
+  *
+  * @param std::string o_chipType - Chip type as output
+  *
+  * @return Upon success, ECMD_SUCCESS will be returned.  A reason code will
+  *         be returned if the execution fails.
+  */
+static uint32_t getChipType(std::string &o_chipType) {
+    int rc=ECMD_SUCCESS;
+    uint32_t l_chip_id = 0;
+    uint32_t l_address = 0x100a; //Bits 12:19 tells chipID
+    ecmdDataBuffer l_data;
+    ecmdChipTarget l_target;
+
+    l_target.cageState =  l_target.nodeState = ECMD_TARGET_FIELD_VALID;
+    l_target.slotState = l_target.posState = ECMD_TARGET_FIELD_VALID;
+
+    //Read the 0x100a cfam to get the chip id
+    rc = dllGetCfamRegister(l_target, l_address, l_data);
+    if (rc){
+        return out.error(EDBG_READ_ERROR, FUNCNAME, "cfam read failure!\n");
+    }
 
-  return rc;
+    //bit 12:19 indicates the chip id and based on that unique id we determine
+    //the chip type.
+    l_chip_id = (l_data.getWord(0) & 0x000ff000) >> 12;
+    switch(l_chip_id) {
+        //0x223'd1'049 - p9
+        case 0xd1:
+            o_chipType = CHIP_TYPE_P9;
+            break;
+
+        default:
+            return out.error(EDBG_GENERAL_ERROR,FUNCNAME, "Unrecognised Chip type" PRIx32 "\n", l_chip_id);
+    }
+    return rc;
 }
 
 // Load the device tree and initialise the targets
 static int initTargets(void) {
-  int fd;
-  void *fdt;
-  struct stat stat;
-  static int done = 0;
-
-  if (!done) {
-    done = 1;
-
-    // The user sets their device tree via this variable. If not set, fail
-    // Would be nice to also set via --device on the cmdline, but currently
-    // there is an order of operations problem.
-    // Longer term libpdbg will be able to auto-detect the correct thing to use.
-    char * devTree = getenv("EDBG_DTB");
-    if (devTree == NULL) {
-      return out.error(ECMD_UNKNOWN_FILE, FUNCNAME, "EDBG_DTB not set in environment, you must set it\n");
-    }
 
-    // If set to 'none', skip the rest of what we do to setup the device tree
-    // This is assuming we won't be using any functions that use the device tree
-    if (!strcmp(devTree, "none")) {
-      return ECMD_SUCCESS;
-    }
-
-    fd = open(devTree, O_RDONLY);
-    if (fd < 0) {
-      return out.error(ECMD_FAILURE, FUNCNAME, "Unable to open device tree: %s\n", devTree);
-    }
-
-    if (fstat(fd, &stat) < 0) {
-      perror("Unable to read device tree size");
-      return ECMD_FAILURE;
-    }
-
-    fdt = mmap(NULL, stat.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
-    if (fdt == MAP_FAILED) {
-      perror("Unable to mmap device tree");
-      return ECMD_FAILURE;
-    }
+  /*  Device tree can also be specified using PDBG_DTB environment variable
+   *  pointing to system device tree.  If system device tree is specified using
+   *  PDBG_DTB, then it will override the default device tree or the specified
+   *  device tree. NULL to use default which is used pdbg. */
+  pdbg_targets_init(NULL);
 
-    pdbg_targets_init(fdt);
+  uint32_t rc = getChipType(chipType);
```

Comment:
Why do we need to do this here?  Why isn't the chipType just checked when it's needed?

Reply:
I can do that. since init is called once in the program I thought it is the best place I can call and assign to the global variable.   I just tried to optimize the calling to the getChipType() every time and instead use the global var instead. But, i can change it if you think it is wrong approach

Reply:
Please have a function that checks the product generation (P9, P10, etc..) that is just called when needed.  That can stay consistent in the code and then call the correct underlying hardware or pdbg function to make that determination.

Reply:
done